br/pkg/storage/locking.go

```diff
@@ -67,58 +174,136 @@ func readLockMeta(ctx context.Context, storage ExternalStorage, path string) (Lo
-	meta, err := readLockMeta(ctx, storage, path)
+func (l RemoteLock) Unlock(ctx context.Context) error {
+	meta, err := readLockMeta(ctx, l.storage, l.path)
 	if err != nil {
 		return err
 	}
 	// NOTE: this is for debug usage. For now, there isn't an Compare-And-Swap
 	// operation in our ExternalStorage abstraction.
 	// So, once our lock has been overwritten or we are overwriting other's lock,
 	// this information will be useful for troubleshooting.
-	log.Info("Releasing lock.", zap.Stringer("meta", meta), zap.String("path", path))
-	err = storage.DeleteFile(ctx, path)
+	if !bytes.Equal(l.txnID[:], meta.TxnID) {
+		return errors.Errorf("Txn ID mismatch: remote is %v, our is %v", meta.TxnID, l.txnID)
+	}
+
+	log.Info("Releasing lock.", zap.Stringer("meta", meta), zap.String("path", l.path))
+	err = l.storage.DeleteFile(ctx, l.path)
 	if err != nil {
-		return errors.Annotatef(err, "failed to delete lock file %s", path)
+		return errors.Annotatef(err, "failed to delete lock file %s", l.path)
 	}
 	return nil
 }
+
+func TryLockRemoteWrite(ctx context.Context, storage ExternalStorage, path, hint string) (lock RemoteLock, err error) {
+	target := fmt.Sprintf("%s.WRIT", path)
+	writer := ExclusiveWrite{
+		Target: target,
+		Content: func(txnID uuid.UUID) []byte {
+			meta := MakeLockMeta(hint)
+			meta.TxnID = txnID[:]
+			res, err := json.Marshal(meta)
+			if err != nil {
+				log.Panic(
+					"Unreachable: a trivial object cannot be marshaled to JSON.",
+					zap.String("path", path),
+					logutil.ShortError(err),
+				)
+			}
+			return res
+		},
+		Verify: func(ctx VerifyWriteContext) error {
+			return ctx.assertNoOtherOfPrefixExpect(path, ctx.IntentFileName())
+		},
+	}
+
+	lock.storage = storage
+	lock.path = target
+	lock.txnID, err = writer.CommitTo(ctx, storage)
+	if err != nil {
+		err = errors.Annotatef(err, "there is something about the lock: %s", tryFetchRemoteLock(ctx, storage, target))
+	}
+	return
+}
+
+func TryLockRemoteRead(ctx context.Context, storage ExternalStorage, path, hint string) (lock RemoteLock, err error) {
+	readID := rand.Int63()
+	target := fmt.Sprintf("%s.READ.%016x", path, readID)
+	writeLock := fmt.Sprintf("%s.WRIT", target)
+	writer := ExclusiveWrite{
+		Target: target,
+		Content: func(txnID uuid.UUID) []byte {
+			meta := MakeLockMeta(hint)
+			meta.TxnID = txnID[:]
+			res, err := json.Marshal(meta)
+			if err != nil {
+				log.Panic(
+					"Unreachable: a trivial object cannot be marshaled to JSON.",
+					zap.String("path", path),
+					logutil.ShortError(err),
+				)
+			}
+			return res
+		},
+		Verify: func(ctx VerifyWriteContext) error {
+			return ctx.assertNoOtherOfPrefixExpect(writeLock, "")
+		},
+	}
+
+	lock.storage = storage
+	lock.path = target
+	lock.txnID, err = writer.CommitTo(ctx, storage)
+	if err != nil {
+		err = errors.Annotatef(err, "there is something about the lock: %s", tryFetchRemoteLock(ctx, storage, writeLock))
```

Comment:
nit, maybe `failed to commit the lock`? I feel like error message should describe what the action was for better debugging.